lib/lte_link_control/lte_lc_helpers.c

```diff
@@ -981,6 +981,340 @@ int parse_ncellmeas(const char *at_response, struct lte_lc_cells_info *cells)
+	 * allocating heap for it. This may save quite a bit of heap as the
+	 * worst case scenario is 96 elements.
+	 * 3 is added to account for the parameters that do not have a trailing
+	 * comma.
+	 */
+	size_t param_count = get_char_frequency(at_response, ',') + 3;
+
+	/* Fill the defaults */
+	cells->gci_cells_count = 0;
+	cells->ncells_count = 0;
+	cells->current_cell.id = LTE_LC_CELL_EUTRAN_ID_INVALID;
+	for (i = 0; i < params->gci_count; i++) {
+		cells->gci_cells[i].id = LTE_LC_CELL_EUTRAN_ID_INVALID;
+		cells->gci_cells[i].timing_advance = LTE_LC_CELL_TIMING_ADVANCE_INVALID;
+	}
+
+	/*
+	 * Response format for GCI search types:
+	 * High level:
+	 * status[,
+	 *	GCI_cell_info1,neighbor_count1[,neighbor_cell1_1,neighbor_cell1_2...],
+	 *	GCI_cell_info2,neighbor_count2[,neighbor_cell2_1,neighbor_cell2_2...]...]
+	 *
+	 * Detailed:
+	 * %NCELLMEAS: status
+	 * [,<cell_id>,<plmn>,<tac>,<ta>,<ta_meas_time>,<earfcn>,<phys_cell_id>,<rsrp>,<rsrq>,
+	 *		<meas_time>,<serving>,<neighbor_count>
+	 *	[,<n_earfcn1>,<n_phys_cell_id1>,<n_rsrp1>,<n_rsrq1>,<time_diff1>]
+	 *	[,<n_earfcn2>,<n_phys_cell_id2>,<n_rsrp2>,<n_rsrq2>,<time_diff2>]...],
+	 *  <cell_id>,<plmn>,<tac>,<ta>,<ta_meas_time>,<earfcn>,<phys_cell_id>,<rsrp>,<rsrq>,
+	 *		<meas_time>,<serving>,<neighbor_count>
+	 *	[,<n_earfcn1>,<n_phys_cell_id1>,<n_rsrp1>,<n_rsrq1>,<time_diff1>]
+	 *	[,<n_earfcn2>,<n_phys_cell_id2>,<n_rsrp2>,<n_rsrq2>,<time_diff2>]...]...
+	 */
+
+	err = at_params_list_init(&resp_list, param_count);
+	if (err) {
+		LOG_ERR("Could not init AT params list for cell_list, error: %d", err);
+		goto clean_exit;
+	}
+
+	err = at_parser_params_from_str(at_response,
+					NULL,
+					&resp_list);
+	if (err && err != -E2BIG) {
+		LOG_ERR("Could not parse AT%%NCELLMEAS response, error: %d", err);
+		goto clean_exit;
+	} else if (err == -E2BIG) {
+		incomplete = true;
+		LOG_WRN("E2BIG was returned, continue. param_count %d, err %d, str %s",
+			param_count, err, at_response);
+	}
+
+	err = at_params_string_get(&resp_list,
+				   AT_RESPONSE_PREFIX_INDEX,
+				   response_prefix,
+				   &response_prefix_len);
+	if (err) {
+		LOG_ERR("Could not get response prefix, error: %d", err);
+		goto clean_exit;
+	}
+
+	if (!response_is_valid(response_prefix, response_prefix_len,
+			       AT_NCELLMEAS_RESPONSE_PREFIX)) {
+		/* The unsolicited response is not a NCELLMEAS response, ignore it. */
+		LOG_ERR("Not a valid NCELLMEAS response");
+		goto clean_exit;
+	}
+
+	/* Status code. */
+	curr_index = AT_NCELLMEAS_STATUS_INDEX;
+	err = at_params_int_get(&resp_list, curr_index, &status);
+	if (err) {
+		LOG_DBG("Cannot parse NCELLMEAS status");
+		goto clean_exit;
+	}
+
+	if (status != AT_NCELLMEAS_STATUS_VALUE_SUCCESS) {
+		err = 1;
+		LOG_DBG("NCELLMEAS status %d", status);
+		goto clean_exit;
+	}
+
+	/* Go through the cells. */
+	for (i = 0; curr_index < (param_count - (AT_NCELLMEAS_GCI_CELL_PARAMS_COUNT + 1)) &&
+			i < params->gci_count; i++) {
+		struct lte_lc_cell parsed_cell;
```

Comment:
Suggestion: It's a pretty long function. Maybe the for-loops could be separated into their own functions and the part for serving cell and its neighbor could be reusable in the `parse_ncellmeas` function.

Reply:
Yeah, I thought that initially as well but ended to this solution, because parse_ncellmeas seems to have little bit different logics and didn't want any change on that code branch.

Reply:
Alright, I guess it can stay as is if there are differences in the logic.
Btw, will you add tests for GCI parsing?

Reply:
OK, I'll add tests in follow-up PR, Jira: NCSDK-17957 created for tracking.